Added get_experiment()

dsblank opened the pull request:
Added ability to create an experiment that is not part of the live, online sequence of experiments created with `create_experiment()`. `get_experiment()` experiments can be created in parallel, and don't close previous experiments.

Example:

```
library("cometr")
exp <- create_experiment()
# This experiment would close previously-created experiments

exp2 <- get_experiment(exp$get_key())
# This experiment does not close any experiments
```
All tests pass.

daattali commented:
Glad to see you're able to run the tests! I still haven't had a chance to take a look sorry, I'm having a crazy week but I'll get around to it as soon as I can.

dsblank commented:
@daattali Thanks for the review.

It sounds like by the end you understand the goal. Yes, "online" is not a good name... I didn't want to use "active" because I think you had already used that term. It is the difference between being in a live, continuous communication with the server, and operating more in a REST fashion.

Do you want to just make these changes you suggested yourself?

dsblank commented on R/config.R:
```diff
@@ -19,6 +19,13 @@ get_config_url <- function() {
   get_config_param("COMET_URL_OVERRIDE", default = .cometrenv$COMET_API_DEFAULT_URL)
 }
 
+#' @importFrom utils head
```
Good to know. Do you know about the github "suggest" feature? You can just click the pages icon above this box.

daattali replied:
I did not, looks useful

dsblank replied:
Fixed

dsblank replied:
Fixed.

dsblank commented on R/config.R:
```diff
@@ -19,6 +19,13 @@ get_config_url <- function() {
+#' @importFrom utils head
+get_config_url_base <- function() {
+  full_url <- get_config_url()
+  expanded <- strsplit(full_url, "/")[[1]]
+  paste0(head(expanded, length(expanded) - 1), "/", collapse="")
```
The get_config_url() has  /clientlib/  on the end of it. This was my code to remove it.

daattali replied:
modify_config_url() removes it


```
modify_config_url <- function(value) {
  sub("/clientlib[/]?$", "", value)
}
```

dsblank replied:
Removed function, used modify_config_url

dsblank replied:
Fixed.

dsblank commented on R/experiment.R:
```diff
@@ -53,6 +53,68 @@ create_experiment <- function(
   keep_active = TRUE, log_output = TRUE, log_error = FALSE,
   log_code = TRUE, log_system_details = TRUE, log_git_info = FALSE
 ) {
+  base_experiment(experiment_name, project_name, workspace_name, api_key,
```
Great place for a github "suggestion".

dsblank replied:
Fixed.

dsblank replied:
Fixed.

dsblank commented on R/experiment.R:
```diff
@@ -53,6 +53,68 @@ create_experiment <- function(
+#'
+#' @export
+get_experiment <- function(
+  experiment_key, api_key = NULL, keep_active = FALSE, log_output = FALSE, log_error = FALSE,
```
Yes, setting to TRUE would get reset (or written for the first time if they weren't before).

dsblank replied:
I think it will be fine to leave this as a base experiment that handles all of the options in one place. That way it leaves maximum flexibility for the user.

dsblank commented on R/experiment.R:
```diff
@@ -76,24 +138,43 @@ create_experiment <- function(
+    workspace_name <- resp[["workspaceName"]]
+    experiment_name <- resp[["experimentName"]]
+    archived <- resp[["archived"]]
+    if (archived) {
```
The link code needs to be written in the next PR anyway, because it should be a dynamic function (depends on the current status of archive for example).

daattali replied:
I'd argue that comet should be smart enough on the server end to do a redirect. If I send someone an experiment's link, and the next day I archive it, the previous link should redirect. Vice versa as well. The current state seems like poor design

dsblank replied:
We have that link too. This URL has more information and is designed for the user to be able to easily see what workspace and project and experiment is in.

dsblank replied:
Moved logic to helper function. Noted the comment about retrieving url in response.

dsblank commented on R/experiment.R:
```diff
@@ -174,22 +256,37 @@ Experiment <- R6::R6Class(
     #' @description
-    #' Do not call this function directly. Use `create_experiment()` instead.
+    #' Do not call this function directly. Use `create_experiment()` or `get_experiment()` instead.
+    #' @param experiment_key The experiment key
```
I was just getting rid of the warnings. Is there a way to suppress warnings? Maybe add a comment then?

daattali replied:
Those warning are only when generating the documents, they're not official R/CRAN warnings. The docs that are generated are still fully valid (in the /man folder). We're using `roxygen2` to generate the documentation, which is not the official way, it's a helper package for generating docs. The warnings are coming from there. If parameters are ommitted from a top-level function, like `get_experiments()`, then you'd get a real warning that would cause the build to fail. These one are safe to ignore. I do have an open issue on github with the roxygen2 package that they shouldn't trigger a warning for this.

dsblank replied:
Removed syntax to generate docs.

dsblank commented on R/experiment.R:
```diff
@@ -174,22 +256,37 @@ Experiment <- R6::R6Class(
+                          keep_active = FALSE, log_output = FALSE, log_error = FALSE,
+			  online = TRUE) {
+      if (isTRUE(online)) {
+        if (!isTRUE(.cometrenv$cancreate)) {
```
I didn't mean for it to be called by users. Perhaps the logic is incorrect, or needs to be duplicated.

dsblank replied:
Good catch! Moved logic.

dsblank commented on R/experiment.R:
```diff
@@ -226,6 +327,12 @@ Experiment <- R6::R6Class(
     },
 
+    #' @description
+    #' Get the online status of an experiment.
```
Yes, please feel free to rename it.

dsblank replied:
Renamed "online" -> "dynamic"

dsblank commented on R/experiment.R:
```diff
@@ -584,7 +692,9 @@ Experiment <- R6::R6Class(
       if (is.null(.cometrenv$curexp) ||
           self$get_key() != .cometrenv$curexp$get_key()) {
-        comet_stop("This experiment already ended and cannot be modified.")
+          if (isTRUE(self$get_online())) {
```
OK

dsblank replied:
Removed test.

dsblank commented on tests/testthat/helper.R:
```diff
@@ -22,8 +22,13 @@ cleanup <- function() {
-  Experiment$new(experiment_key = experiment_key, api_key = test_api_key,
-                 keep_active = keep_active, log_output = log_output, log_error = log_error)
+  with_mock(
+    `cometr:::log_other` = function(experiment_key, api_key, key, value) {
```
I couldn't really understand mocks well enough to do better.

daattali replied:
You used mocks correctly. My argument here is that the underlying code (not the testing code) should not call log_other directly. The experiment creation should call a different function to make the call to log_other.

daattali replied:
If you look at some of my mock functions, when I need to use the values then I actually name them, but when all I need to do is silence a function then I use `...`. It's not an official guideline, just something I thought would be safer.

dsblank replied:
Fixed.

dsblank commented on tests/testthat/helper.R:
```diff
@@ -22,8 +22,13 @@ cleanup <- function() {
-                 keep_active = keep_active, log_output = log_output, log_error = log_error)
+  with_mock(
+    `cometr:::log_other` = function(experiment_key, api_key, key, value) {
+      # Keep new experiments from logging "Created by", "cometr"
```
I tried to make as few as changes as needed in this PR.

daattali replied:
It would only involve changing `log_other(key=..., value=...)` to `log_initial_values <- function() { log_other(key=..., value=...)}`. It would solve the problem that I mention here, and it would also create self-documenting code so that you won't need the comment that you added here ("Keep new experiments from ...")

dsblank replied:
Got it. Thanks! Fixed.

daattali commented:
I'd rather not tinker with the code of someone else's PR, you can implement the changes

dsblank commented on R/experiment.R:
```diff
@@ -76,24 +138,43 @@ create_experiment <- function(
+    LOG_INFO("Experiment created: ", experiment_link, echo = TRUE)
+    .cometrenv$cancreate <- TRUE
+  } else {
+    online <- FALSE
```
Changed "online" to "dynamic"

dsblank commented on R/experiment.R:
```diff
@@ -53,6 +53,68 @@ create_experiment <- function(
+  experiment_name = NULL, project_name = NULL, workspace_name = NULL, api_key = NULL,
+  keep_active = TRUE, log_output = TRUE, log_error = FALSE,
+  log_code = TRUE, log_system_details = TRUE, log_git_info = FALSE,
+  experiment_key = NULL
```
Good catch!  Added check

dsblank commented on R/experiment.R:
```diff
@@ -174,22 +256,37 @@ Experiment <- R6::R6Class(
-      if (!isTRUE(.cometrenv$cancreate)) {
-        comet_stop("Do not call this function directly. Use `create_experiment()` instead.")
+                          keep_active = FALSE, log_output = FALSE, log_error = FALSE,
+			  online = TRUE) {
```
Fixed.

dsblank commented on R/experiment.R:
```diff
@@ -174,22 +256,37 @@ Experiment <- R6::R6Class(
+        }
+        LOG_DEBUG("Creating experiment ", experiment_key)
+
+          .cometrenv$cancreate <- FALSE
```
Fixed

dsblank commented on R/experiment.R:
```diff
@@ -174,22 +256,37 @@ Experiment <- R6::R6Class(
+        LOG_DEBUG("Creating experiment ", experiment_key)
+
+          .cometrenv$cancreate <- FALSE
+	  .cometrenv$curexp <- self
```
Fixed.

dsblank commented on R/experiment.R:
```diff
@@ -218,6 +315,10 @@ Experiment <- R6::R6Class(
       }
+
+      if (isTRUE(online)) {
+	self$log_other(key = "Created by", value = "cometr")
```
Fixed.

dsblank commented on NAMESPACE:
```diff
@@ -1,13 +1,13 @@
 # Generated by roxygen2: do not edit by hand
 
-export(Experiment)
```
This probably should not have been included originally?

daattali replied:
I assume you removed this line yourself? The NAMESPACE file should never be touched. It's auto generated and should be left alone. This needs to be included, otherwise a lot of documentation is missing.

dsblank replied:
No, that would not be a valid assumption. That must have been removed by the auto documentation system.

daattali replied:
Then something got broken somehow. Because that line should definitely still be in the namespace. Right now the entire experiment object won't have any documentation

dsblank replied:
Apparently the "@description" causes that to be added, so I added that back.

dsblank commented:
Updated PR. Major points:

dsblank commented:
Updated PR. Major points:

* renamed "online" to "dynamic" (but open to better name)
* fixed mock to just a single function
* fixed spacing issues
* addressed all comments
* we can add tests for the new functions in another PR
* goal: don't break the existing tests, release next version

dsblank commented:
Travis still fails, with a cut off log: https://travis-ci.org/github/comet-ml/cometr/builds/682601086

My local check passes.

daattali commented on R/experiment.R:
```diff
@@ -137,9 +139,12 @@ base_experiment <- function(
   if (log_git_info && (!requireNamespace("git2r", quietly = TRUE) || utils::packageVersion("git2r") < "0.22.1")) {
     comet_stop("log_git_info requires you to have `git2r` version 0.22.1 or later.")
   }
+  if (!is.null(experiment_key) && (!is.null(workspace_name) || !is.null(project_name))) {
```
The order of operations here is unclear. Is it    `A & (B | C)` or `(A & B) | C)`? The way it's currently written, if I was to look at this line of code next week, I would not know what the correct logic here is.

daattali commented on R/experiment.R:
```diff
@@ -137,9 +139,12 @@ base_experiment <- function(
   if (log_git_info && (!requireNamespace("git2r", quietly = TRUE) || utils::packageVersion("git2r") < "0.22.1")) {
     comet_stop("log_git_info requires you to have `git2r` version 0.22.1 or later.")
   }
+  if (!is.null(experiment_key) && (!is.null(workspace_name) || !is.null(project_name))) {
```
So experiment_name is ok?

dsblank replied:
I don't understand this question. Ca you explain?

daattali replied:
My original comment was that experiment_name or workspace or project - none of them should be provided. The check here is for only two of those. So experiment_name and expreiment_key can both be provided

dsblank replied:
Ok, thanks! I added it, and adjusted the text to be more clear.

daattali commented on R/experiment.R:
```diff
@@ -137,9 +139,12 @@ base_experiment <- function(
     comet_stop("log_git_info requires you to have `git2r` version 0.22.1 or later.")
   }
+  if (!is.null(experiment_key) && (!is.null(workspace_name) || !is.null(project_name))) {
+    comet_stop("either experiment_key or (workspace_name and project_name) must be given, not both.")
```
I would capitalize the sentence, to keep in line with the other messages, and look more professional

dsblank replied:
```suggestion
    comet_stop("Either experiment_key or (workspace_name and project_name) must be given, not both.")
```

daattali commented on R/experiment.R:
```diff
@@ -316,8 +317,8 @@ Experiment <- R6::R6Class(
-      if (isTRUE(online)) {
-	self$log_other(key = "Created by", value = "cometr")
+      if (isTRUE(dynamic)) {
+        initialize_dynamic_experiment(self)
```
I would choose a different name. `initialize_dynamic_experiment` sets the expectation that this function is going to do some initialization. Which is false. All it does it log metadata information, so its name should reflect that.

dsblank replied:
There may be more processing done in the future to initial these experiments.

daattali replied:
Whatever processing might happen, you'll likely want that processing to happen in testing as well, no? It seemed to me that adding the metadata was a special case of something you don't want to happen. If you think any other processing is also going to be excluded from tests, then this name does make sense.

dsblank replied:
Renamed `initialize_dynamic_experiment` to `experiment_log_metadata`

daattali commented on R/utils.R:
```diff
@@ -17,3 +17,29 @@ get_values_from_list <- function(list, key) {
+    )
+  } else {
+    experiment_link <- paste(base_url,
+                             workspace_name,
```
The previous lines are all indented by 6 chars, but these are are indented by.... a lot.

dsblank replied:
I used the default R studio settings.

daattali replied:
Start the first parameter on a new line, then if you copy-paste the rest of the lines they'll follow suit

dsblank replied:
I think in a follow-up PR I'll format all of the code to use the R studio standard.

dsblank replied:
Fixed

daattali commented on R/experiment.R:
```diff
@@ -158,21 +163,18 @@ base_experiment <- function(
-      experiment_link <- paste(get_config_url_base(),
-        workspace_name, "/", project_name, "/", experiment_key, sep="")
-    }
+    experiment_link = create_experiment_link(modify_config_url(get_config_url()),
```
Similar to a previous comment I made, when calling a function with more than 3-4 parameters, it's a very good idea to name them

dsblank replied:
Fixed!

daattali commented:
Rstudio is not opinionated about style

On Sun., May 3, 2020, 13:12 Doug Blank, <notifications@github.com> wrote:

> *@dsblank* commented on this pull request.
> ------------------------------
>
> In R/utils.R
> <https://github.com/comet-ml/cometr/pull/16#discussion_r419132158>:
>
> > +}
> +
> +create_experiment_link <- function(base_url, workspace_name, project_name, experiment_key, archived) {
> +  if (archived) {
> +    experiment_link <- paste(
> +      base_url,
> +      workspace_name,
> +      "/",
> +      project_name,
> +      "/archived/",
> +      experiment_key,
> +      sep = ""
> +    )
> +  } else {
> +    experiment_link <- paste(base_url,
> +                             workspace_name,
>
> I think in a follow-up PR I'll format all of the code to use the R studio
> standard.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/comet-ml/cometr/pull/16#discussion_r419132158>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHIQFAL5CWYUL7S7TYE4XLRPWQWZANCNFSM4MQIJJWQ>
> .
>

daattali commented on R/experiment.R:
```diff
@@ -60,6 +60,8 @@ create_experiment <- function(
     log_code = log_code, log_system_details = log_system_details, log_git_info = log_git_info)
 }
 
+#' @title get_experiment
```
By specifying it this way, the title of the function in documentaiton will be "get_experiment" instead of "Get a previously created experiment". You shouldn't need to use @title and @description

dsblank replied:
I added these to get it to show up in the docs, to solve cross-reference issues. 

I'm still trying to figure out why Travis tests aren't passing... grasping at straws...

daattali replied:
I'm not sure what failure the tests are getting, but I'd suggest using the correct title and description here

dsblank replied:
The tests are now passing! This must have been the cause. I'll change the title to your suggestion.

dsblank replied:
Tests pass! https://travis-ci.org/github/comet-ml/cometr/builds/683897091

daattali commented:
You can use explicit @title and @description tags, but just use the correct
title if you do. With roxygen, if these tags aren't provided, it takes the
first line and treats that as title, and the next paragraph is desc. Either
way is fine

On Wed., May 6, 2020, 11:49 Doug Blank, <notifications@github.com> wrote:

> *@dsblank* commented on this pull request.
> ------------------------------
>
> In R/experiment.R
> <https://github.com/comet-ml/cometr/pull/16#discussion_r420898264>:
>
> > @@ -60,6 +60,8 @@ create_experiment <- function(
>      log_code = log_code, log_system_details = log_system_details, log_git_info = log_git_info)
>  }
>
> +#' @title get_experiment
>
> The tests are now passing! This must have been the cause. I'll change the
> title to your suggestion.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/comet-ml/cometr/pull/16#discussion_r420898264>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHIQFD3BCDR245UL7QC62TRQGBHHANCNFSM4MQIJJWQ>
> .
>

dsblank commented:
> With roxygen, if these tags aren't provided, it takes the first line and treats that as title, and the next paragraph is desc.

Of course that is what I had, and it didn't work (Travis tests did not pass). Perhaps a difference in some code that you are using versus what Travis is using? In any event, the previous commit passed. Checking to see if this one passes...

dsblank commented:
https://travis-ci.org/github/comet-ml/cometr/builds/683897091

daattali commented:
Maybe you didn't run the documentation generation last time before running
the tests?

On Wed., May 6, 2020, 11:58 Doug Blank, <notifications@github.com> wrote:

> https://travis-ci.org/github/comet-ml/cometr/builds/683897091
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/comet-ml/cometr/pull/16#issuecomment-624735431>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHIQFHYRJEHUX2DZSPIPL3RQGCJZANCNFSM4MQIJJWQ>
> .
>

dsblank commented:
> Maybe you didn't run the documentation generation last time before running the tests?

No, I think the new generated page was just missing from the repo. I think that was it the whole time. In any event, I learned a lot about R development through this.

dsblank commented:
I believe that this is ready for final review.
